### ISSUE TITLE
No error when `__enter__` has extra arguments

### ISSUE THREAD

**Issue by JelleZijlstra** [microsoft/pyright#2874]
This will fail at runtime but pyright is OK with it:
```python
class BadCM2:
    def __enter__(self, extra_arg) -> int:
        return 0

    def __exit__(self, typ, value, tb):
        pass

def capybara():
    with BadCM2():
        pass
```

**Comment by JelleZijlstra** [comment]
(And the same happens with `async with`)

**Comment by erictraut** [comment]
I'm on the fence on this one. Does this fall within the bounds of a type checker's responsibilities? Arguably, but it's a slippery slope because it means that pyright would need to get into the business of validating all magic method signatures.

**Comment by JelleZijlstra** [comment]
For what it's worth, in pyanalyze I type check `with` statements by seeing if the with target is compatible with a Protocol that defines `__enter__` and `__exit__` appropriately, so it will naturally produce an error about the signature of `__enter__`. The error would be produced on the `with` line, not the definition of `__enter__`.

**Comment by erictraut** [comment]
I don't think this is a common error. Mypy doesn't implement it either, and I can't find a single request for the feature in the mypy issue tracker.

For now, I'm going to close this. I might consider adding it in the future if I receive requests from others or this gets upvoted (thumbs up).
